### ISSUE TITLE
GUI setup of two factor authentication

### ISSUE THREAD

**Issue by Justinzobel** [cockpit-project/cockpit#15860]
Page: 

A feature within the web interface to setup and enable 2FA for users would be really good.

Cockpit is really powerful and in that way extra security should be available without extra workarounds.

**Comment by martinpitt** [comment]
You can set up 2FA with e.g. the [PAM Google Authenticator](https://wiki.archlinux.org/title/Google_Authenticator) (for a single machine) or centrally if you use identity, e.g. with [FreeIPA](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/linux_domain_identity_authentication_and_policy_guide/otp). There should not be any workarounds necessary for Cockpit, it uses the normal global PAM authentication settings.

You can restrict 2FA to cockpit logins only if you configure it in /etc/pam.d/cockpit instead of /etc/pam.d/system-auth (or the equivalent Debian/Ubuntu file).

**Comment by Justinzobel** [comment]
A GUI setup would be nice to add. So I can log in to cockpit once my server is up, enable 2FA and get a QR code to scan on the web page without having to touch a terminal.

**Comment by CodebijSven** [comment]
Is this feature coming?

**Comment by jelly** [comment]
The Cockpit Team is not actively working on this feature, we welcome anyone to contribute this feature. So if you have any questions feel free to ask them.

**Comment by brolifen** [comment]
I hope someone picks this up as cyber security is no joke. A lot of people that use cockpit have limited Linux knowledge in the first place, so to provide this as an out of box UI option would not only make cockpit more secure but would secure a lot of environments with limited Linux skill sets.

**Comment by poVoq** [comment]
Looks like this was implemented for the nethserver fork: https://community.nethserver.org/t/2fa-or-two-factor-authentication-with-cockpit/14172

Would be cool if that could be back-ported to vanilla cockpit.

**Comment by jelly** [comment]
> Looks like this was implemented for the nethserver fork: https://community.nethserver.org/t/2fa-or-two-factor-authentication-with-cockpit/14172
> 
> Would be cool if that could be back-ported to vanilla cockpit.

If you use 2FA with Cockpit via PAM I believe this should just work out of the box. Having this documented might be nice.

**Comment by brolifen** [comment]
> > Looks like this was implemented for the nethserver fork: https://community.nethserver.org/t/2fa-or-two-factor-authentication-with-cockpit/14172
> > Would be cool if that could be back-ported to vanilla cockpit.
> 
> If you use 2FA with Cockpit via PAM I believe this should just work out of the box. Having this documented might be nice.

It does indeed work but a GUI option would be very nice. For the people that come across this issue and want to enable MFA right now here is a one liner:

`sudo dnf install google-authenticator qrencode-libs -y;google-authenticator -t -d -f -r 3 -R 30 -W -Q UTF8; sudo bash -c 'echo "auth required pam_google_authenticator.so nullok" >> /etc/pam.d/cockpit';sudo systemctl restart cockpit;

Please someone make this a GUI option

**Comment by JaneX8** [comment]
> It does indeed work but a GUI option would be very nice. For the people that come across this issue and want to enable MFA right now here is a one liner:
> 
> `sudo dnf install google-authenticator qrencode-libs -y;google-authenticator -t -d -f -r 3 -R 30 -W -Q UTF8; sudo bash -c 'echo "auth required pam_google_authenticator.so nullok" >> /etc/pam.d/cockpit';sudo systemctl restart cockpit;`
> 
> Please someone make this a GUI option

I agree that a GUI option is needed and thanks for pointing out this config file. However, why are you using `nullok` in the end? Doesn't that mean `if no config, ignore this PAM module`, so perhaps accidentally bypassing 2FA? I think it is safer to use `auth required pam_google_authenticator.so` (without nullok) instead. Or am I missing something?

**Comment by shr0048** [comment]
@JaneX8 

Great guide for who need 2-fa auth!

Here for Debian / Ubuntu users version

`
sudo apt-get install libpam-google-authenticator libqrencode-dev -y
`

`
google-authenticator -t -d -f -r 3 -R 30 -W -Q UTF8
`

`
sudo bash -c 'echo "auth required pam_google_authenticator.so nullok" >> /etc/pam.d/cockpit'
`

`sudo systemctl restart cockpit
`

save back-up code and emergency scratch codes somewhere safe!

**Comment by GuilhermeLimaSP** [comment]
Cockpit doesn't have native 2fa GUI or brute force protection? I'm scared

**Comment by poVoq** [comment]
It has and it works great.

This topic is about making it easier to set up through the GUI.

**Comment by Justinzobel** [comment]
> hi, i installed today cockpit on debian 12, and want to enable 2fa, using this guide:
> 
> ```
> sudo apt-get install libpam-google-authenticator libqrencode-dev -y
> 
> google-authenticator -t -d -f -r 3 -R 30 -W -Q UTF8
> 
> sudo bash -c 'echo "auth required pam_google_authenticator.so nullok" >> /etc/pam.d/cockpit'
> 
> sudo systemctl restart cockpit 
> ```
> 
> but 2fa code is not requested on login to non root users!, if i delete `nullok` from `/etc/pam.d/cockpit` 2fa login is requested on non root users, but OTP code is not valid, how i can enable 2fa authentication to another user that root? best regards

This isn't related to getting 2FA added into the GUI. Please create a new issue.

**Comment by ggsammyy** [comment]
I'd love to see a GUI to get this done.

**Comment by CasperOng** [comment]
sooo is yubikey login (passkey login) / 2fa via fido2 key still haven't implemented?

**Comment by jelly** [comment]
> sooo is yubikey login (passkey login) / 2fa via fido2 key still haven't implemented?

There is development on going for passkey support it will take a while before it is complete https://github.com/cockpit-project/cockpit/pull/22689

**Comment by Justinzobel** [comment]
> There is development on going for passkey support it will take a while before it is complete [#22689](https://github.com/cockpit-project/cockpit/pull/22689)

This isn't relevant to this issue.

**Comment by Venefilyn** [comment]
On the topic of GUI specifically for TOTP/HOTP setup, what modules do people use? Is it still a majority of people using `pam_oath`?

**Comment by Anticept** [comment]
> You can set up 2FA with e.g. the [PAM Google Authenticator](https://wiki.archlinux.org/title/Google_Authenticator) (for a single machine) or centrally if you use identity, e.g. with [FreeIPA](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/linux_domain_identity_authentication_and_policy_guide/otp). There should not be any workarounds necessary for Cockpit, it uses the normal global PAM authentication settings.
> 
> You can restrict 2FA to cockpit logins only if you configure it in /etc/pam.d/cockpit instead of /etc/pam.d/system-auth (or the equivalent Debian/Ubuntu file).

As a FreeIPA user: this does not work. If the "send hardened password via spake or fast" is allowed, it will use this, and bypass the OTP. If you DISABLE that option, you cannot log in at all regardless of your password+OTP allow/disallow.

For that matter, disabling spake/fast also breaks password+otp at the web logon to the freeipa management interface.

running debian 13, cockpit 337-1+deb13u1, sssd 2.10.1-2+b1, freeipa is on almalinux 10.1, ipa-server: 4.13.1 Release : 3.el10_2.2
